### PR TITLE
notify: add markdown support for wechat

### PR DESCRIPTION
**Pull request by deamwork**
According to the WeChat work official API document, adding markdown type message support.

https://translate.google.com/translate?hl=en&sl=zh-CN&tl=en&u=https%3A%2F%2Fwork.weixin.qq.com%2Fapi%2Fdoc%2F90000%2F90135%2F90236%23markdown%25E6%25B6%2588%25E6%2581%25AF&prev=search

### CODE REVIEW

File: config/notifiers.go [hunk 1]
```diff
@@ -396,14 +396,15 @@ type WechatConfig struct {
...
-	AgentID   string `yaml:"agent_id,omitempty" json:"agent_id,omitempty"`
+	APISecret   Secret `yaml:"api_secret,omitempty" json:"api_secret,omitempty"`
+	CorpID      string `yaml:"corp_id,omitempty" json:"corp_id,omitempty"`
+	Message     string `yaml:"message,omitempty" json:"message,omitempty"`
+	APIURL      *URL   `yaml:"api_url,omitempty" json:"api_url,omitempty"`
+	ToUser      string `yaml:"to_user,omitempty" json:"to_user,omitempty"`
+	ToParty     string `yaml:"to_party,omitempty" json:"to_party,omitempty"`
+	ToTag       string `yaml:"to_tag,omitempty" json:"to_tag,omitempty"`
+	AgentID     string `yaml:"agent_id,omitempty" json:"agent_id,omitempty"`
+	MessageType string `yaml:"message_type,omitempty" json:"message_type,omitempty"`
```
Comment (simonpasquier) [inline config/notifiers.go:407]: please modify the `*WechatConfig.UnmarshalYAML()` method to default to `text` if empty and check that the value is either `text` or `markdown`.

File: notify/wechat/wechat.go [hunk 2]
```diff
@@ -134,16 +135,29 @@ func (n *Notifier) Notify(ctx context.Context, as ...*types.Alert) (bool, error)
 		n.accessTokenAt = time.Now()
 	}
 
+	// Define base message params
```
Comment (simonpasquier) [inline notify/wechat/wechat.go:138]: no need for a comment here

File: notify/wechat/wechat.go [hunk 3]
```diff
@@ -134,16 +135,29 @@ func (n *Notifier) Notify(ctx context.Context, as ...*types.Alert) (bool, error)
...
-		},
 		ToUser:  tmpl(n.conf.ToUser),
 		ToParty: tmpl(n.conf.ToParty),
 		Totag:   tmpl(n.conf.ToTag),
 		AgentID: tmpl(n.conf.AgentID),
-		Type:    "text",
-		Safe:    "0",
+	}
+
+	// Determine message type
```
Comment (simonpasquier) [inline notify/wechat/wechat.go:146]: same here, you can remove the comment

File: notify/wechat/wechat.go [hunk 4]
```diff
@@ -134,16 +135,29 @@ func (n *Notifier) Notify(ctx context.Context, as ...*types.Alert) (bool, error)
...
+	switch n.conf.MessageType {
```
Comment (simonpasquier) [inline notify/wechat/wechat.go:147]: IMO it's simpler to write it like this:

```go
if msg.Type == "markdown" {
    msg.Markdown = weChatMessageContent{
        Content: tmpl(n.conf.Message),
    }
} else {
    msg.Text = weChatMessageContent{
        Content: tmpl(n.conf.Message),
    }
}
```

File: notify/wechat/wechat.go [hunk 5]
```diff
@@ -134,16 +135,29 @@ func (n *Notifier) Notify(ctx context.Context, as ...*types.Alert) (bool, error)
...
+	// Define base message params
 	msg := &weChatMessage{
-		Text: weChatMessageContent{
-			Content: tmpl(n.conf.Message),
-		},
 		ToUser:  tmpl(n.conf.ToUser),
 		ToParty: tmpl(n.conf.ToParty),
 		Totag:   tmpl(n.conf.ToTag),
 		AgentID: tmpl(n.conf.AgentID),
-		Type:    "text",
```
Comment (simonpasquier) [inline notify/wechat/wechat.go:145]: You can keep this line as well as the next one

```go
    Type = n.conf.MessageType,
    Safe: "0",
```

Comment (deamwork) [reply notify/wechat/wechat.go:145]: According to the document, `markdown` mode doesn't support `safe` param.
So I move this param for `text` mode specifically.

File: docs/configuration.md [hunk 6]
```diff
@@ -711,6 +711,8 @@ API](http://admin.wechat.com/wiki/index.php?title=Customer_Service_Messages).
 
 # API request data as defined by the WeChat API.
 [ message: <tmpl_string> | default = '{{ template "wechat.default.message" . }}' ]
+# The message type only support for `text` and `markdown`.
```
Comment (simonpasquier) [inline docs/configuration.md:714]:
```suggestion
# Type of the message type, supported values are `text` and `markdown`.
```

File: config/notifiers.go [hunk 7]
```diff
@@ -396,21 +396,38 @@ type WechatConfig struct {
...
+	if err := unmarshal((*plain)(c)); err != nil {
+		return err
+	}
+
+	if c.MessageType == "" {
+		c.MessageType = "text"
+	}
+
+	if !wechatTypeMatcher.MatchString(c.MessageType) {
+		return errors.Errorf("WeChat responder `MessageType` type does not match valid options %s", wechatValidTypesRe)
```
Comment (simonpasquier) [inline config/notifiers.go:427]:
```suggestion
		return errors.Errorf("WeChat message type %q does not match valid options %s", c.MessageType, wechatValidTypesRe)
```

File: config/notifiers.go [hunk 8]
```diff
@@ -396,21 +396,38 @@ type WechatConfig struct {
...
+	Message     string `yaml:"message,omitempty" json:"message,omitempty"`
+	APIURL      *URL   `yaml:"api_url,omitempty" json:"api_url,omitempty"`
+	ToUser      string `yaml:"to_user,omitempty" json:"to_user,omitempty"`
+	ToParty     string `yaml:"to_party,omitempty" json:"to_party,omitempty"`
+	ToTag       string `yaml:"to_tag,omitempty" json:"to_tag,omitempty"`
+	AgentID     string `yaml:"agent_id,omitempty" json:"agent_id,omitempty"`
+	MessageType string `yaml:"message_type,omitempty" json:"message_type,omitempty"`
 }
 
+const wechatValidTypesRe = `^(text|markdown)$`
```
Comment (simonpasquier) [inline config/notifiers.go:410]: It would be worth adding a test in `config/notifiers_test.go`. Check `TestOpsgenieTypeMatcher` for inspiration.

**Comment by deamwork** [comment]
And a little question here:
 Is this project a CNCF project?

**Comment by simonpasquier** [comment]
@deamwork yes it is

**Pull request merged**